src/crimson/osd/state.h

```diff
@@ -43,32 +51,47 @@ class OSDState {
                        : wait_for_active.get_shared_future();
   };
   bool is_prestop() const {
+    ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
     return state == State::PRESTOP;
   }
   bool is_stopping() const {
     return state == State::STOPPING;
   }
   bool is_waiting_for_healthy() const {
+    ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
     return state == State::WAITING_FOR_HEALTHY;
   }
   void set_preboot() {
+    ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
     state = State::PREBOOT;
   }
   void set_booting() {
+    ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
     state = State::BOOTING;
   }
   void set_active() {
```

Comment:
This implementation suggests that set_active() can be called from shards other than PRIMARY_CORE, but doing so would only update that shard.  It's also problematic that it returns before actually updating all shards.  Let's split this up as follows:

```
  /// Sets local instance state to active, called from set_active
  void _set_active() {
    state = State::ACTIVE;
    wait_for_active.set_value();
    wait_for_active = {};
  }
public:
  ...
  /// Sets all shards to active
  seastar::future<> set_active() {
      ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
      return container().invoke_on_all([](auto& osd_state) {
        osd_state._set_active();
      });
  }
```

Reply:
done.